src/JsonService.ts

```diff
@@ -176,6 +187,15 @@ export class JsonService {
         if (!response.ok) {
             logger.error("Error from server:", json);
             if (json.error) {
+                if (json.error === "use_dpop_nonce" && dpopEnabled && response.headers.get("Dpop-Nonce")) {
```

Comment:
The spec [describes an optional nonce requirement](https://datatracker.ietf.org/doc/html/rfc9449#name-resource-server-provided-no) whereby if a nonce is not supplied during a request, a bad request is returned by the resource server with a specific response body and the client is expected to generate a nonce and retry the request.

Reply:
lets keep the throw of that error, if the IDP needs a `Dpop-Nonce`, we should already provide one in the first place...

Reply:
Yeah, I'm okay with that but it's worth considering that the spec is a little unclear about when an authorization server should require a dpop_nonce (keep in mind that the nonce is generated server side):

>Including a nonce value contributed by the authorization server in the DPoP proof MAY be used by authorization servers to limit the lifetime of DPoP proofs. The server determines when to issue a new DPoP nonce challenge and if it is needed, thereby requiring the use of the nonce value in subsequent DPoP proofs. The logic through which the server makes that determination is out of scope of this document.

but it's quite clear about what to do when no nonce is supplied:

>An authorization server MAY supply a nonce value to be included by the client in DPoP proofs sent. In this case, the authorization server responds to requests that do not include a nonce with an HTTP 400 (Bad Request) error response per [Section 5.2](https://rfc-editor.org/rfc/rfc6749#section-5.2) of [[RFC6749](https://datatracker.ietf.org/doc/html/rfc6749)] using use_dpop_nonce as the error code value. The authorization server includes a DPoP-Nonce HTTP header in the response supplying a nonce value to be used when sending the subsequent request. Nonce values MUST be unpredictable. This same error code is used when supplying a new nonce value when there was a nonce mismatch. The client will typically retry the request with the new nonce value supplied upon receiving a use_dpop_nonce error with an accompanying nonce value.

This is how it's implemented in IdentityServer for example - the nonce header isn't returned anywhere in the flow, prior to responding with a 400 bad request if the drop_nonce header doesn't exist. This is also how [Okta](https://developer.okta.com/docs/guides/dpop/main/#oauth-2-0-dpop-jwt-flow) does it. So, happy to remove it, but I suspect what is here will become the norm if it hasn't already.

Reply:
Hey @pamapa I have removed the authorization server supplied nonce support for the time being. I think it would be better to do it in a separate PR if that's okay (it is [optional](https://datatracker.ietf.org/doc/html/rfc9449#section-8) in the spec)? This PR is large and it was becoming messy.